engine/crates/federated-graph/src/from_sdl.rs

```diff
@@ -479,46 +470,49 @@ fn ingest_entity_keys(parsed: &ast::ServiceDocument, state: &mut State<'_>) -> R
     Ok(())
 }
 
-fn ingest_field_directives_after_graph(
-    parsed: &ast::ServiceDocument,
-    state: &mut State<'_>,
+fn ingest_field_directives_after_graph<'a>(
+    parsed: &'a ast::TypeSystemDocument,
+    state: &mut State<'a>,
 ) -> Result<(), DomainError> {
-    for definition in &parsed.definitions {
-        let ast::TypeSystemDefinition::Type(typedef) = definition else {
+    for definition in parsed.definitions() {
+        let ast::Definition::Type(typedef) = definition else {
             continue;
         };
-        let fields = match &typedef.node.kind {
-            ast::TypeKind::Object(object) => &object.fields,
-            ast::TypeKind::Interface(iface) => &iface.fields,
+
+        let fields = match typedef {
+            ast::TypeDefinition::Object(object) => object.fields(),
+            ast::TypeDefinition::Interface(iface) => iface.fields(),
             _ => continue,
         };
 
-        let parent_id = state.definition_names[typedef.node.name.node.as_str()];
-        ingest_join_field_directive(parent_id, &typedef.node.directives, fields, state)?;
+        let parent_id = state.definition_names[typedef.name()];
+
+        ingest_join_field_directive(parent_id, || typedef.directives(), fields, state)?;
         ingest_authorized_directive(parent_id, fields, state)?;
     }
 
     Ok(())
 }
 
-fn ingest_join_field_directive(
+fn ingest_join_field_directive<'a, I>(
     parent_id: Definition,
-    parent_directives: &[Positioned<ConstDirective>],
-    fields: &[Positioned<FieldDefinition>],
+    parent_directives: impl Fn() -> I,
+    fields: impl Iterator<Item = ast::FieldDefinition<'a>>,
     state: &mut State<'_>,
-) -> Result<(), DomainError> {
-    let is_federated_entity = parent_directives
-        .iter()
-        .filter(|dir| dir.node.name.node == JOIN_TYPE_DIRECTIVE_NAME)
-        .any(|dir| dir.node.get_argument("key").is_some());
+) -> Result<(), DomainError>
+where
+    I: Iterator<Item = ast::Directive<'a>>,
+{
+    let is_federated_entity = parent_directives()
+        .filter(|dir| dir.name() == JOIN_TYPE_DIRECTIVE_NAME)
+        .any(|dir| dir.get_argument("key").is_some());
 
     let type_subgraph_ids = if !is_federated_entity {
-        parent_directives
-            .iter()
-            .filter(|dir| dir.node.name.node == JOIN_TYPE_DIRECTIVE_NAME)
+        parent_directives()
```

Comment:
I may be missing some reason why this wouldn't work, but iirc I made `Iter: Clone + Copy` so I expect parent_directives could just be an `ast::iter::Iter<'a, ast::Directive<'a>>`

Reply:
It looks like this particular instance doesn't implement Clone

Reply:
ah well, damn

Reply:
Fixed that here: https://github.com/obmarg/cynic/pull/1030